rocAL/rocAL/source/pipeline/tensor.cpp

```diff
@@ -0,0 +1,357 @@
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+*/
+
+#include <cstdio>
+#if !ENABLE_HIP
+#include <CL/cl.h>
+#endif
+#include <stdexcept>
+#include <vx_ext_amd.h>
+#include <cstring>
+#include "commons.h"
+#include "tensor.h"
+
+vx_enum vx_mem_type(RocalMemType mem)
+{
+    switch(mem)
+    {
+        case RocalMemType::OCL:
+        {
+            return VX_MEMORY_TYPE_OPENCL;
+        }
+        break;
+        case RocalMemType::HOST:
+        {
+            return VX_MEMORY_TYPE_HOST;
+        }
+        break;
+        case RocalMemType::HIP:
+        {
+            return VX_MEMORY_TYPE_HIP;
+        }
+        break;
+        default:
+            throw std::runtime_error("Memory type not valid");
+    }
+}
+
+vx_size tensor_data_size(RocalTensorDataType data_type)
+{
+    switch(data_type)
+    {
+        case RocalTensorDataType::FP32:
+        {
+            return sizeof(vx_float32);
+        }
+        break;
+        case RocalTensorDataType::FP16:
+        {
+            return sizeof(vx_int16);
+        }
+        break;
+        case RocalTensorDataType::UINT8:
+        {
+            return sizeof(vx_uint8);
+        }
+        case RocalTensorDataType::UINT32:
+        {
+            return sizeof(vx_uint32);
+        }
+        case RocalTensorDataType::INT32:
+        {
+            return sizeof(vx_int32);
+        }
+        break;
+        default:
+            throw std::runtime_error("tensor data_type not valid");
+    }
+}
+
+bool operator==(const rocALTensorInfo &rhs, const rocALTensorInfo &lhs)
+{
+    return (rhs.dims() == lhs.dims() &&
+            rhs.mem_type() == lhs.mem_type() &&
+            rhs.data_type() == lhs.data_type() &&
+            rhs.color_format() == lhs.color_format() &&
+            rhs.layout() == lhs.layout());
+}
+
+void rocALTensorInfo::reallocate_tensor_roi_buffers()
+{
+    _roi = std::make_shared<std::vector<RocalROI>>(_batch_size);
+
+    if(_roi->size() > 0)
+        _roi->clear();
+    _roi->resize(_batch_size);
+    if(_is_image)
```

Comment:
I don't think, we need to check this here. Please do a memset() to 0 instead which will be much more efficient